library/psa_crypto_pake.c

```diff
@@ -190,272 +164,116 @@ static psa_status_t mbedtls_ecjpake_to_psa_error(int ret)
-        status = PSA_ERROR_BAD_STATE;
-        goto error;
-    }
-
-    if (user_id_len == 0 || user_id == NULL) {
-        status = PSA_ERROR_INVALID_ARGUMENT;
         goto error;
     }
 
-    status = PSA_ERROR_NOT_SUPPORTED;
-
-error:
-    psa_pake_abort(operation);
-    return status;
-}
-
-psa_status_t psa_pake_set_peer(psa_pake_operation_t *operation,
-                               const uint8_t *peer_id,
-                               size_t peer_id_len)
-{
-    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
-
-    if (operation->alg == PSA_ALG_NONE ||
-        operation->state != PSA_PAKE_STATE_SETUP) {
-        status = PSA_ERROR_BAD_STATE;
-        goto error;
-    }
-
-    if (peer_id_len == 0 || peer_id == NULL) {
-        status = PSA_ERROR_INVALID_ARGUMENT;
-        goto error;
-    }
-
-    status = PSA_ERROR_NOT_SUPPORTED;
-
-error:
-    psa_pake_abort(operation);
-    return status;
-}
-
-psa_status_t psa_pake_set_role(psa_pake_operation_t *operation,
-                               psa_pake_role_t role)
-{
-    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
-
-    if (operation->alg == PSA_ALG_NONE ||
-        operation->state != PSA_PAKE_STATE_SETUP) {
-        status = PSA_ERROR_BAD_STATE;
-        goto error;
-    }
-
-    if (role != PSA_PAKE_ROLE_NONE &&
-        role != PSA_PAKE_ROLE_FIRST &&
-        role != PSA_PAKE_ROLE_SECOND &&
-        role != PSA_PAKE_ROLE_CLIENT &&
-        role != PSA_PAKE_ROLE_SERVER) {
-        status = PSA_ERROR_INVALID_ARGUMENT;
-        goto error;
-    }
+    operation->password_len = actual_password_len;
+    operation->alg = cipher_suite.algorithm;
 
 #if defined(MBEDTLS_PSA_BUILTIN_ALG_JPAKE)
-    if (operation->alg == PSA_ALG_JPAKE) {
-        if (role != PSA_PAKE_ROLE_CLIENT &&
-            role != PSA_PAKE_ROLE_SERVER) {
-            return PSA_ERROR_NOT_SUPPORTED;
+    if (cipher_suite.algorithm == PSA_ALG_JPAKE) {
+        if (cipher_suite.type != PSA_PAKE_PRIMITIVE_TYPE_ECC ||
+            cipher_suite.family != PSA_ECC_FAMILY_SECP_R1 ||
+            cipher_suite.bits != 256 ||
+            cipher_suite.hash != PSA_ALG_SHA_256) {
+            status = PSA_ERROR_NOT_SUPPORTED;
+            goto error;
         }
 
         operation->role = role;
 
+        operation->buffer_length = 0;
+        operation->buffer_offset = 0;
+
+        status = psa_pake_ecjpake_setup(operation);
+        if (status != PSA_SUCCESS) {
+            goto error;
+        }
+
         return PSA_SUCCESS;
     } else
+#else
+    (void) operation;
+    (void) inputs;
 #endif
-    status = PSA_ERROR_NOT_SUPPORTED;
+    { status = PSA_ERROR_NOT_SUPPORTED; }
 
 error:
-    psa_pake_abort(operation);
+    /* When driver fails with PSA_ERROR_NOT_SUPPORTED the built-in implementation is executed (if available)
+       and it will reallocate the password leading to the memory leak.
+       Call abort explicitly to clean up allocated memory for password on failure. */
```

Comment:
If built-in implementation also fails, then the core will call `mbedtls_psa_pake_abort()`, but it will release password allocated by built-in implementation and password allocated by driver is a memory leak.
The built-in implementation may also succeed and then we will also have a memory leak on the driver side.

Reply:
The better place to call the driver pake abort seems the driver wrapper as that's the place where we know which driver returned in error and that we are going to use the same memory space to call the next driver or builtin implementation. In `psa_driver_wrapper_pake_setup` it would be something like:

```C
    switch( location )                                                          
    {                                                                           
        case PSA_KEY_LOCATION_LOCAL_STORAGE:                                    
            /* Key is stored in the slot in export representation, so           
             * cycle through all known transparent accelerators */              
#if defined(PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT)                              
#if defined(PSA_CRYPTO_DRIVER_TEST)                                             
            status = mbedtls_test_transparent_pake_setup(                       
                        &operation->data.ctx.transparent_test_driver_ctx,       
                        inputs );                                               
            if( status == PSA_SUCCESS )
                operation->id = MBEDTLS_TEST_TRANSPARENT_DRIVER_ID;
            else
                mbedtls_test_transparent_pake_abort(&operation->data.ctx.transparent_test_driver_ctx);
            /* Declared with fallback == true */                                
            if( status != PSA_ERROR_NOT_SUPPORTED )                             
                return( status );                                               
#endif /* PSA_CRYPTO_DRIVER_TEST */                                             
#endif /* PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT */                              
#if defined(MBEDTLS_PSA_BUILTIN_PAKE)                                           
            status = mbedtls_psa_pake_setup( &operation->data.ctx.mbedtls_ctx,  
                        inputs );                                               
            if( status == PSA_SUCCESS )                                         
                operation->id = PSA_CRYPTO_MBED_TLS_DRIVER_ID;
           else
                mbedtls_psa_pake_abort(&operation->data.ctx.mbedtls_ctx);                
            return status;                                                      
#endif                                                                          
            return( PSA_ERROR_NOT_SUPPORTED );                                  
        /* Add cases for opaque driver here */                                  
        default:                                                                
            /* Key is declared with a lifetime not known to us */               
            (void)operation;                                                    
            (void)inputs;                                                       
            (void)status;                                                       
            return( PSA_ERROR_INVALID_ARGUMENT );                               
    }
```

Reply:
I disagree, I don't think this belongs in the driver wrapper. The driver wrapper is generated code. It should keep as simple as possible. It's the core's job to guarantee that `abort` is called on error paths. It would do that anyway to clean up the core-level data structures.

Reply:
This seems better solution as the diver doesn't care about this special case for `setup` (required `abort` call on failure).

Reply:
> I disagree, I don't think this belongs in the driver wrapper. The driver wrapper is generated code. It should keep as simple as possible. It's the core's job to guarantee that abort is called on error paths. It would do that anyway to clean up the core-level data structures.

We have the following case:
- `setup` driver entry point allocates memory for password and then fails with not supported result.
- driver wrapper tries next driver or built-in implementation which also allocates new memory for password and we have memory leak. We didn't reach the core level yet.

In current version setup driver entry point calls `abort` on failure to release memory.

Reply:
That looks like a bug in the first driver's setup. It's supposed to clean after itself on error.

Quoting the [driver interface specification](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/proposed/psa-driver-interface.md#general-considerations-on-multi-part-operations):

> The core calls the xxx_setup entry point for this operation family. If this fails, the core destroys the operation context object without calling any other driver entry point on it.

Reply:
For entry points other then `setup` the failure is propagated to core and core calls `abort` and the wrapper calls abort for the specific driver (based on `operation->id`). 
In case of `setup` situation is different as many drivers setup entry points can be executed to find supported one. So I think that in this case each setup driver entry point should clean up after itself on failure.